File: HADES/plugins/patchright/hades_bridge.py
#!/usr/bin/env python3
"""HADES bridge for Patchright (stealth Playwright drop-in)."""
from __future__ import annotations

import argparse
import ipaddress
import json
import socket
import subprocess
import sys
from pathlib import Path
from urllib.parse import urlparse


def doctor() -> dict:
    import cli_bridge

    payload = cli_bridge.doctor(["patchright"], ["patchright"])
    payload["notes"] = [
        "Run install_browsers once after dependency install.",
        "Use fetch/screenshot for research pages that block vanilla Playwright.",
        "This is browser automation — not an exploit framework.",
    ]
    return payload


def install_browsers() -> dict:
    process = subprocess.run(
        [sys.executable, "-m", "patchright", "install", "chromium"],
        capture_output=True,
        text=True,
        timeout=1800,
        shell=False,
    )
    exit_code = int(process.returncode or 0)
    return {
        "ok": exit_code == 0,
        "command": [sys.executable, "-m", "patchright", "install", "chromium"],
        "exit_code": exit_code,
        "stdout": (process.stdout or "")[-40_000:],
        "stderr": (process.stderr or "")[-40_000:],
        **({"error": "browser_install_failed"} if exit_code != 0 else {}),
    }


def _launch():
    from patchright.sync_api import sync_playwright

    return sync_playwright()


def _resolved_addresses(host: str, port: int) -> set[ipaddress.IPv4Address | ipaddress.IPv6Address]:
    try:
        return {ipaddress.ip_address(host)}
    except ValueError:
        pass
    try:
        infos = socket.getaddrinfo(host, port, type=socket.SOCK_STREAM)
    except OSError as exc:
        raise ValueError(f"Host resolution failed: {host}") from exc
    addresses: set[ipaddress.IPv4Address | ipaddress.IPv6Address] = set()
    for info in infos:
        raw = str(info[4][0]).split("%", 1)[0]
        try:
            addresses.add(ipaddress.ip_address(raw))
        except ValueError:
            continue
    if not addresses:
        raise ValueError(f"Host resolved to no usable addresses: {host}")
    return addresses


def ensure_public_url(url: str) -> str:
    parsed = urlparse(str(url).strip())
    if parsed.scheme.lower() not in {"http", "https"}:
        raise ValueError("Only http/https URLs are allowed")
    host = (parsed.hostname or "").lower().strip("[]")
    if not host or host in {"localhost", "metadata.google.internal"} or host.endswith((".localhost", ".local", ".internal")):
        raise ValueError(f"Private/local host blocked: {host or '<empty>'}")
    port = parsed.port or (443 if parsed.scheme.lower() == "https" else 80)
    addresses = _resolved_addresses(host, port)
    blocked = [str(address) for address in addresses if not address.is_global]
    if blocked:
        raise ValueError(f"Private/non-public address blocked for {host}: {', '.join(sorted(blocked))}")
    return parsed.geturl()


def _install_request_guard(page) -> None:
    def guard(route) -> None:
        request_url = str(route.request.url or "")
        parsed = urlparse(request_url)
        if parsed.scheme.lower() not in {"http", "https"}:
            route.continue_()
            return
        try:
            ensure_public_url(request_url)
        except ValueError:
            route.abort("blockedbyclient")
            return
        route.continue_()

    page.route("**/*", guard)


def _blocked_result(url: str, exc: Exception, *, output: str | None = None) -> dict:
    payload = {"url": url, "ok": False, "status": 0, "error": str(exc)}
    if output is None:
        payload.update({"chars": 0, "text": ""})
    else:
        payload.update({"path": "", "output": output, "bytes": 0})
    return payload


def fetch(url: str, max_chars: int) -> dict:
    try:
        url = ensure_public_url(url)
    except ValueError as exc:
        return _blocked_result(url, exc)

    with _launch() as p:
        browser = p.chromium.launch(headless=True)
        try:
            page = browser.new_page()
            _install_request_guard(page)
            response = page.goto(url, wait_until="domcontentloaded", timeout=60_000)
            try:
                ensure_public_url(page.url)
            except ValueError as exc:
                return _blocked_result(url, exc)
            status = int(response.status) if response is not None else 0
            if response is None or status >= 400:
                return {
                    "url": url,
                    "ok": False,
                    "status": status,
                    "error": f"HTTP {status or 'no response'}",
                    "chars": 0,
                    "text": "",
                }
            title = page.title()
            text = page.inner_text("body") or ""
            return {
                "url": page.url,
                "title": title,
                "ok": bool(str(text).strip()),
                "status": status,
                "chars": len(text),
                "truncated": len(text) > max_chars,
                "text": text[:max_chars],
                **({} if str(text).strip() else {"error": "empty fetch result"}),
            }
        finally:
            browser.close()


def screenshot(url: str, output: str) -> dict:
    try:
        url = ensure_public_url(url)
    except ValueError as exc:
        return _blocked_result(url, exc, output=output)

    out = Path(output)
    out.parent.mkdir(parents=True, exist_ok=True)
    with _launch() as p:
        browser = p.chromium.launch(headless=True)
        try:
            page = browser.new_page(viewport={"width": 1280, "height": 720})
            _install_request_guard(page)
            response = page.goto(url, wait_until="networkidle", timeout=60_000)
            try:
                ensure_public_url(page.url)
            except ValueError as exc:
                return _blocked_result(url, exc, output=str(out.resolve()))
            status = int(response.status) if response is not None else 0
            if response is None or status >= 400:
                return {
                    "url": url,
                    "ok": False,
                    "status": status,
                    "error": f"HTTP {status or 'no response'}",
                    "output": str(out.resolve()),
                    "bytes": 0,
                }
            page.screenshot(path=str(out), full_page=True)
            bytes_written = out.stat().st_size if out.exists() else 0
            return {
                "url": page.url,
                "ok": bytes_written > 0,
                "status": status,
                "output": str(out.resolve()),
                "bytes": bytes_written,
                **({} if bytes_written > 0 else {"error": "empty screenshot"}),
            }
        finally:
            browser.close()


def main() -> int:
    parser = argparse.ArgumentParser()
    sub = parser.add_subparsers(dest="cmd", required=True)
    sub.add_parser("doctor")
    sub.add_parser("install_browsers")
    f = sub.add_parser("fetch")
    f.add_argument("--url", required=True)
    f.add_argument("--max-chars", type=int, default=20000)
    s = sub.add_parser("screenshot")
    s.add_argument("--url", required=True)
    s.add_argument("--output", default="screenshot.png")
    args = parser.parse_args()
    if args.cmd == "doctor":
        payload = doctor()
        print(json.dumps(payload, ensure_ascii=False, indent=2))
        return 0 if payload.get("ok", True) else 2
    if args.cmd == "install_browsers":
        payload = install_browsers()
        print(json.dumps(payload, ensure_ascii=False, indent=2))
        return 0 if payload.get("ok") else 1
    if args.cmd == "fetch":
        payload = fetch(args.url, args.max_chars)
        print(json.dumps(payload, ensure_ascii=False, indent=2))
        if payload.get("ok") is False:
            return 1
        if not str(payload.get("text") or "").strip():
            return 2
        return 0
    payload = screenshot(args.url, args.output)
    print(json.dumps(payload, ensure_ascii=False, indent=2))
    if payload.get("ok") is False:
        return 1
    if int(payload.get("bytes") or 0) <= 0:
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: HADES/plugins/puppeteer/hades_bridge.mjs
import dns from "node:dns/promises";
import fs from "node:fs";
import net from "node:net";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const args = process.argv.slice(2);
const cmd = args[0];

function argValue(flag, fallback = "") {
  const idx = args.indexOf(flag);
  if (idx === -1) return fallback;
  const value = args[idx + 1];
  if (!value || value.startsWith("--")) return fallback;
  return value;
}

async function withBrowser(fn) {
  const puppeteer = await import("puppeteer");
  // Do not disable Chromium's process sandbox. If a host cannot launch the
  // sandbox, surface that as an operational error instead of silently dropping
  // a core browser security boundary.
  const browser = await puppeteer.default.launch({ headless: true });
  try {
    return await fn(browser);
  } finally {
    await browser.close();
  }
}

function blockedIpv4(address) {
  const parts = address.split(".").map(Number);
  if (parts.length !== 4 || parts.some((part) => !Number.isInteger(part) || part < 0 || part > 255)) return true;
  const [a, b, c] = parts;
  if (a === 0 || a === 10 || a === 127) return true;
  if (a === 100 && b >= 64 && b <= 127) return true;
  if (a === 169 && b === 254) return true;
  if (a === 172 && b >= 16 && b <= 31) return true;
  if (a === 192 && b === 0 && c === 0) return true;
  if (a === 192 && b === 0 && c === 2) return true;
  if (a === 192 && b === 88 && c === 99) return true;
  if (a === 192 && b === 168) return true;
  if (a === 198 && (b === 18 || b === 19)) return true;
  if (a === 198 && b === 51 && c === 100) return true;
  if (a === 203 && b === 0 && c === 113) return true;
  if (a >= 224) return true;
  return false;
}

function blockedIpv6(address) {
  const value = address.toLowerCase().split("%", 1)[0];
  if (value === "::" || value === "::1") return true;
  if (value.startsWith("::ffff:")) {
    const mapped = value.slice("::ffff:".length);
    return net.isIP(mapped) !== 4 || blockedIpv4(mapped);
  }

  // Be conservative: ordinary globally routed IPv6 unicast lives in 2000::/3.
  // Reject deprecated IPv4-compatible forms (::x.x.x.x), ULA/link-local,
  // multicast and other non-global families by rejecting anything outside it.
  const pieces = value.split(":");
  const first = Number.parseInt(pieces[0] || "0", 16);
  const second = Number.parseInt(pieces[1] || "0", 16);
  if (!Number.isFinite(first) || !Number.isFinite(second)) return true;
  if (first < 0x2000 || first > 0x3fff) return true;

  // Transition/benchmark/documentation address space must not become a route
  // around the public-only browser boundary. 2001:0..2f includes IETF special
  // assignments such as Teredo/benchmark/ORCHID; normal public 2001 prefixes
  // (e.g. 2001:4860::) remain available.
  if (first === 0x2001 && second <= 0x002f) return true;
  if (first === 0x2001 && second === 0x0db8) return true; // documentation
  if (first === 0x2002) return true; // deprecated 6to4 embeds an IPv4 route
  if (first === 0x3fff) return true; // documentation/special-use allocation
  return false;
}

function blockedAddress(address) {
  const family = net.isIP(address);
  if (family === 4) return blockedIpv4(address);
  if (family === 6) return blockedIpv6(address);
  return true;
}

async function ensurePublicUrl(rawUrl) {
  let parsed;
  try {
    parsed = new URL(String(rawUrl || "").trim());
  } catch {
    throw new Error("Invalid URL");
  }
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    throw new Error("Only http/https URLs are allowed");
  }
  const host = (parsed.hostname || "").toLowerCase().replace(/^\[|\]$/g, "");
  if (!host || host === "localhost" || host === "metadata.google.internal" || host.endsWith(".localhost") || host.endsWith(".local") || host.endsWith(".internal")) {
    throw new Error(`Private/local host blocked: ${host || "<empty>"}`);
  }

  let addresses;
  if (net.isIP(host)) {
    addresses = [{ address: host }];
  } else {
    try {
      addresses = await dns.lookup(host, { all: true, verbatim: true });
    } catch (error) {
      throw new Error(`Host resolution failed: ${host}`, { cause: error });
    }
  }
  if (!addresses.length) throw new Error(`Host resolved to no usable addresses: ${host}`);
  const blocked = addresses.map((item) => String(item.address || "")).filter((address) => blockedAddress(address));
  if (blocked.length) {
    throw new Error(`Private/non-public address blocked for ${host}: ${blocked.sort().join(", ")}`);
  }
  return parsed.toString();
}

async function installRequestGuard(page) {
  await page.setRequestInterception(true);
  page.on("request", async (request) => {
    const requestUrl = request.url();
    let parsed;
    try {
      parsed = new URL(requestUrl);
    } catch {
      try { await request.abort("blockedbyclient"); } catch {}
      return;
    }
    if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
      try { await request.continue(); } catch {}
      return;
    }
    try {
      await ensurePublicUrl(requestUrl);
      await request.continue();
    } catch {
      try { await request.abort("blockedbyclient"); } catch {}
    }
  });
}

async function guardedGoto(page, url, options) {
  const safeUrl = await ensurePublicUrl(url);
  await installRequestGuard(page);
  const response = await page.goto(safeUrl, options);
  const finalUrl = await ensurePublicUrl(page.url());
  return { response, finalUrl };
}

if (cmd === "doctor") {
  const pkgPath = path.join(__dirname, "package.json");
  const pkg = JSON.parse(fs.readFileSync(pkgPath, "utf8"));
  let puppeteerOk = false;
  try {
    await import("puppeteer");
    puppeteerOk = true;
  } catch {
    puppeteerOk = false;
  }
  const payload = {
    ok: puppeteerOk,
    name: pkg.name,
    version: pkg.version,
    dependency: pkg.dependencies?.puppeteer || null,
    puppeteer_importable: puppeteerOk,
  };
  if (!puppeteerOk) payload.error = "puppeteer_not_importable";
  console.log(JSON.stringify(payload, null, 2));
  process.exit(puppeteerOk ? 0 : 2);
}

if (cmd === "fetch") {
  const url = argValue("--url");
  const maxChars = Number(argValue("--max-chars", "20000"));
  if (!url) {
    console.error("url is required");
    process.exit(1);
  }
  try {
    await ensurePublicUrl(url);
  } catch (error) {
    console.log(JSON.stringify({ url, ok: false, error: String(error?.message || error), status: 0 }, null, 2));
    process.exit(2);
  }
  const result = await withBrowser(async (browser) => {
    const page = await browser.newPage();
    let response;
    let finalUrl;
    try {
      ({ response, finalUrl } = await guardedGoto(page, url, { waitUntil: "domcontentloaded", timeout: 60000 }));
    } catch (error) {
      return { url, ok: false, error: String(error?.message || error), status: 0 };
    }
    const status = response ? response.status() : 0;
    if (!response || status >= 400) {
      return { url: finalUrl || url, ok: false, error: `HTTP ${status || "no response"}`, status };
    }
    const title = await page.title();
    const text = await page.evaluate(() => document.body?.innerText || "");
    const html = await page.content();
    return {
      url: finalUrl,
      title,
      ok: Boolean(String(text || "").trim()),
      status,
      chars: text.length,
      truncated: text.length > maxChars,
      text: text.slice(0, maxChars),
      html_chars: html.length,
      ...(String(text || "").trim() ? {} : { error: "empty page text" }),
    };
  });
  console.log(JSON.stringify(result, null, 2));
  if (result.ok === false) process.exit(1);
  if (!String(result.text || "").trim()) process.exit(2);
  process.exit(0);
}

if (cmd === "screenshot") {
  const url = argValue("--url");
  const output = argValue("--output", "screenshot.png");
  if (!url) {
    console.error("url is required");
    process.exit(1);
  }
  try {
    await ensurePublicUrl(url);
  } catch (error) {
    console.log(JSON.stringify({ url, ok: false, error: String(error?.message || error), status: 0, bytes: 0 }, null, 2));
    process.exit(2);
  }
  const result = await withBrowser(async (browser) => {
    const page = await browser.newPage();
    await page.setViewport({ width: 1280, height: 720 });
    let response;
    let finalUrl;
    try {
      ({ response, finalUrl } = await guardedGoto(page, url, { waitUntil: "networkidle2", timeout: 60000 }));
    } catch (error) {
      return { url, ok: false, error: String(error?.message || error), status: 0, bytes: 0 };
    }
    const status = response ? response.status() : 0;
    if (!response || status >= 400) {
      return { url: finalUrl || url, ok: false, error: `HTTP ${status || "no response"}`, status, bytes: 0 };
    }
    const outPath = path.resolve(output);
    await page.screenshot({ path: outPath, fullPage: true });
    const bytes = fs.statSync(outPath).size;
    return {
      url: finalUrl,
      output: outPath,
      bytes,
      status,
      ok: bytes > 0,
      ...(bytes > 0 ? {} : { error: "empty screenshot" }),
    };
  });
  console.log(JSON.stringify(result, null, 2));
  if (result.ok === false) process.exit(1);
  if (!result.bytes) process.exit(2);
  process.exit(0);
}

console.error(`unknown command: ${cmd}`);
process.exit(1);
